dbms/src/DataStreams/MultiplexInputStream.h

```diff
@@ -0,0 +1,248 @@
+// limitations under the License.
+
+#pragma once
+
+#include <DataStreams/IProfilingBlockInputStream.h>
+
+#include <memory>
+#include <queue>
+#include <vector>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+} // namespace ErrorCodes
+
+class MultiPartitionStreamPool
+{
+public:
+    MultiPartitionStreamPool() = default;
+
+    void addPartitionStreams(const BlockInputStreams & cur_streams)
+    {
+        if (cur_streams.empty())
+            return;
+        std::unique_lock lk(mu);
+        streams_queue_by_partition.push_back(
+            std::make_shared<std::queue<std::shared_ptr<IBlockInputStream>>>());
+        for (const auto & stream : cur_streams)
+            streams_queue_by_partition.back()->push(stream);
+        added_streams.insert(added_streams.end(), cur_streams.begin(), cur_streams.end());
+    }
+
+    std::shared_ptr<IBlockInputStream> pickOne()
+    {
+        std::unique_lock lk(mu);
+        if (streams_queue_by_partition.empty())
+            return nullptr;
+        if (streams_queue_id >= static_cast<int>(streams_queue_by_partition.size()))
+            streams_queue_id = 0;
+
+        auto & q = *streams_queue_by_partition[streams_queue_id];
+        std::shared_ptr<IBlockInputStream> ret = nullptr;
+        assert(!q.empty());
+        ret = q.front();
+        q.pop();
+        if (q.empty())
+            streams_queue_id = removeQueue(streams_queue_id);
+        else
+            streams_queue_id = nextQueueId(streams_queue_id);
+        return ret;
+    }
+
+    int exportAddedStreams(BlockInputStreams & ret_streams)
+    {
+        std::unique_lock lk(mu);
+        for (auto & stream : added_streams)
+            ret_streams.push_back(stream);
+        return added_streams.size();
+    }
+
+    int addedStreamsCnt()
+    {
+        std::unique_lock lk(mu);
+        return added_streams.size();
+    }
+
+private:
+    int removeQueue(int queue_id)
+    {
+        streams_queue_by_partition[queue_id] = nullptr;
+        if (queue_id != static_cast<int>(streams_queue_by_partition.size()) - 1)
+        {
+            swap(streams_queue_by_partition[queue_id], streams_queue_by_partition.back());
+            streams_queue_by_partition.pop_back();
+            return queue_id;
+        }
+        else
+        {
+            streams_queue_by_partition.pop_back();
+            return 0;
+        }
+    }
+
+    int nextQueueId(int queue_id) const
+    {
+        if (queue_id + 1 < static_cast<int>(streams_queue_by_partition.size()))
+            return queue_id + 1;
+        else
+            return 0;
+    }
+
+    static void swap(std::shared_ptr<std::queue<std::shared_ptr<IBlockInputStream>>> & a,
+                     std::shared_ptr<std::queue<std::shared_ptr<IBlockInputStream>>> & b)
+    {
+        auto tmp = a;
+        a = b;
+        b = tmp;
+    }
+
+    std::vector<
+        std::shared_ptr<std::queue<
+            std::shared_ptr<IBlockInputStream>>>>
```

Comment:
nit: according to ClickHouse's coding style, better to add an alias for `std::queue<std::shared_ptr<IBlockInputStream>>` like:

```cpp
using BlockInputStreamPtrQueue = std::queue<std::shared_ptr<IBlockInputStream>>;
```